test/elixir/test/cookie_auth_test.exs

```diff
@@ -0,0 +1,401 @@
+    jason_user_doc_rev = resp.body["rev"]
+
+    # can't delete another users doc unless you are admin
+
+    jchris_user_doc = Map.put(jchris_user_doc, "_rev", jchris_rev)
+
+    delete_as(
+      @users_db,
+      jchris_user_doc,
+      use_session: session,
+      expect_response: 404,
+      error_message: "not_found"
+    )
+
+    logout(session)
+
+    # test redirect on success
+    resp =
+      Couch.post(
+        "/_session",
+        query: [next: "/_up"],
+        body: %{
+          :username => "Jason Davies",
+          :password => @password
+        }
+      )
+
+    assert resp.status_code == 302
+    assert resp.body["ok"]
+    assert String.ends_with?(resp.headers["location"], "/_up")
+
+    # test redirect on fail
+    resp =
+      Couch.post(
+        "/_session",
+        query: [fail: "/_up"],
+        body: %{
+          :username => "Jason Davies",
+          :password => "foobar"
+        }
+      )
+
+    assert resp.status_code == 302
+    assert resp.body["error"] == "unauthorized"
+    assert String.ends_with?(resp.headers["location"], "/_up")
+
+    session = login("jchris", "funnybone")
+    info = Couch.Session.info(session)
+    assert info["userCtx"]["name"] == "jchris"
+    assert Enum.empty?(info["userCtx"]["roles"])
+
+    jason_user_doc =
+      jason_user_doc
+      |> Map.put("_rev", jason_user_doc_rev)
+      |> Map.put("foo", 3)
+
+    save_as(
+      @users_db,
+      jason_user_doc,
+      use_session: session,
+      expect_response: 404,
+      error_message: "not_found"
+    )
+
+    jchris_user_doc = Map.put(jchris_user_doc, "roles", ["foo"])
+
+    save_as(
+      @users_db,
+      jchris_user_doc,
+      use_session: session,
+      expect_response: 403,
+      error_message: "forbidden"
+    )
+
+    logout(session)
+
+    jchris_user_doc = Map.put(jchris_user_doc, "foo", ["foo"])
+
+    resp =
+      save_as(
+        @users_db,
+        jchris_user_doc,
+        user: "jan"
+      )
+
+    # wait for auth cache invalidation
+    :timer.sleep(500)
+
+    # test that you can't save system (underscore) roles even if you are admin
+    jchris_user_doc =
+      jchris_user_doc
+      |> Map.put("roles", ["_bar"])
+      |> Map.put("_rev", resp.body["rev"])
+
+    save_as(
+      @users_db,
+      jchris_user_doc,
+      user: "jan",
+      expect_response: 403,
+      error_message: "forbidden"
+    )
+
+    session = login("jchris", "funnybone")
+    info = Couch.Session.info(session)
+
+    assert not Enum.member?(info["userCtx"]["roles"], "_admin")
+    assert(Enum.member?(info["userCtx"]["roles"], "foo"))
+
+    logout(session)
+
+    login("jan", "apple")
+
+    run_on_modified_server(
+      [
+        %{
+          :section => "admins",
+          :key => "jchris",
+          :value => "funnybone"
+        }
+      ],
+      &test_change_admin_fun/0
+    )
+
+    # log in one last time so run_on_modified_server can clean up the admin account
+    login("jan", "apple")
+  after
```

Comment:
This is usually a job for test teardown. In ExUnit it is accomplished via [`on_exit`](https://hexdocs.pm/ex_unit/ExUnit.Callbacks.html#on_exit/2)